### DISCUSSION TITLE
What is the thinking behind records not requiring all properties to be initialized, yet complaining about non-nullables?

### DISCUSSION THREAD

**Discussion by reinux** [dotnet/csharplang#4857] [General]
Coming from F#, this doesn't make much sense to me:

```
public record User
{
  public string FirstName { get; init; } // warning CS8618: Non-nullable property 'FirstName' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. 
  public string LastName { get; init; } // warning CS8618: Non-nullable property 'LastName' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. 
  public string EmailAddress { get; init; } // warning CS8618: Non-nullable property 'EmailAddress' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. 
}

...

new User  // no errors, no warnings
{
}
```

It seems that if you have non-nullable fields, the only way to make this right is to use a constructor, which then disallows object initialization syntax, and brings it half way back to just being a class with readonly fields again (except it allows using with to do updates).

In F#, you receive an error if in the object initialization you don't specify all fields. That way there's no reason for it to warn about unassigned non-nulls.

What's the reasoning behind this weird, seemingly self-defeating behavior? Am I doing something wrong?

Might it make sense to remove the CS8618 warning for record fields when the record has no constructor, and instead warn if in the object initialization, a (non-nullable) property isn't assigned to?

**Comment by acaly** [comment]
I guess once we have the required initializer this should be solved.

#3630

**Comment by Joe4evr** [comment]
> What's the reasoning behind this weird, seemingly self-defeating behavior?

The reasoning is that *declaring* a property to be non-null and *initialization* at the construction site are two orthogonal concerns.

**Comment by DavidArno** [comment]
It is because `init` is a simple compiler "trick" addition to properties. [If you view the IL of the following](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIDMABAE4CmAxgPYnpEBKAsAFADeLRRhXAjAAxEAGkVZEA5mQQBuIrCjSiAXyItFQA==):
```cs
public record R
{
  public string X { get; init; } 
}
```
you'll see that it still declares a `set_X` method as it would do if you used `public string X { get; set; }`, but the compiler then prevents you calling that setter except during construction and initialisation. 

The C# team take a iterative approach to adding such features. As @acaly says, the next step in the process is to add some sort of "required properties" feature:
```cs
public record User
{
    public string FirstName { get; required init; } 
    public string LastName { get; required init; }
    public string EmailAddress { get; required init; }
}
```
At which point, you should no longer get those `null` warnings and you should get errors with:
```cs
new User { }
```

If this seems a back to front approach, welcome to the whacky world of C# where "hills of success" are favoured over "pits of success" all too often. 🤷

**Comment by HaloFour** [comment]
TLDR: records and NRTs are 100% orthogonal.  NRTs are both optional and designed very intentionally to have no effect on the behavior of the code other than to generate diagnostics.  Records need to compile and behave identically with or without NRTs enabled.

Required properties should alleviate some of the issue here as that should move the warning to the site of consumption.  Required properties would give you a warning if you do not set the property, and NRTs, if enabled, would give you a warning if you set the property to a possibly-null reference.  The compiler could then drop the diagnostic at the declaration site.

**Reply by DavidArno** [reply]
I may be misunderstanding the OP , but I do not believe the issue is with records as such. It's the fact that `init` properties are not required to be initialised by default. The fact that I can create a non-nullable reference init property and the compiler complains it's not initialised by the constructor, but doesn't comment that it's not initialised by an initialisation block that is the "WUT?" here.

**Reply by HaloFour** [reply]
Right.  Init properties are still orthogonal to NRTs, and they don't specify that a property must be initialized, only when, so they still don't cover this case.  Required properties cover that.

**Reply by reinux** [reply]
I think functional programming features are effective specifically because they impose artificial constraints that allow for useful assumptions. All of those constraints need to be satisfied in normal usage, even if they depend on features that are orthogonal to each other.

One of the key characteristic of records and immutability in general is that it encourages objects to be in a valid state at all times, beginning at instantiation. I'm not sure how it does that if it's warning you not to use NRTs.

The use cases for an immutable data structure that doesn't guarantee that everything is initialized up-front are pretty limited. As it stands, it's effectively just syntactic sugar for deep (ish) copies and value equality.

I have to agree with @DavidArno that unless there's some kind of technical reason that it _had_ to be done this way, it very much feels like a feature that was rushed out the door without much foresight. Even with the `required` keyword coming, the language has kind of dug itself into an awkward situation of having to recommending users not to use records without it.

**Reply by acaly** [reply]
> even if they depend on features that are orthogonal to each other.

I agree that even if two features are orthogonal, as long as they coexist in the same language, they need to be consistent to some extent. After all, since most features are designed separately, we may say most of them are orthogonal. That's not an excuse at all. If one specific feature in C# encourages user to pay more attention to nullables, there should not be another feature (even being orthogonal) that itself weakens the nullable safety.

The real reason behind this is, as I said, there is a pending feature that will finally link the two together.

**Reply by acaly** [reply]
Honestly, I have tries multiple times in my own project to adopt the nullable and records but yet have no real success. Nullable usually ends up with some error or warning that I don't bother fix (I think many have no good fix yet because of pending feature, etc.). Records as you said usually become effectively the same as normal classes as I add new methods. I know Microsoft is using these features in many big projects already, but until these features become more mature, maybe smaller projects don't get too much considering the efforts to rewrite related codes.

**Reply by theunrepentantgeek** [reply]
> maybe smaller projects don't get too much considering the efforts to rewrite related codes.

I disagree. 

I've migrated multiple small projects to use nullable and in all of them it's been worth the minor effort required. In each I've been able to get them to compile cleanly - without warnings - and in most the compiler's flow analysis has found at least one edge case where things might have gone boom 💥. [I blogged about one example](http://nichesoftware.co.nz/2019/09/28/wordtutor-dotnet30-csharp8.html), look for the heading _Fields not initialized at construction_.

**Reply by reinux** [reply]
Positional records are also nigh useless for Entity Framework or anything requiring field attributes, as it simply doesn't allow them.

I honestly think this implementation of records was a mistake.

**Reply by CyrusNajmabadi** [reply]
I don't see the connection you're making here. You are conflating null vs initialized. I think it's a mistake to do that. Null can be totally legal for something. And whether or not something is initialized is an orthogonal concept.

**Reply by reinux** [reply]
@CyrusNajmabadi Please see [my comment above](https://github.com/dotnet/csharplang/discussions/4857#discussioncomment-908056) on why these two orthogonal concepts are necessary for a useful implementation of immutability and therefore of records.

**Reply by acaly** [reply]
> I don't see the connection you're making here. You are conflating null vs initialized. I think it's a mistake to do that. Null can be totally legal for something. And whether or not something is initialized is an orthogonal concept.

You know, orthogonal vectors can become not so when projected. These two features are related because if you don't initialize something, that thing is null. Currently without "required members", you simply cannot properly initialize a non-nullable variable without violating the contract. I understand the language team designed and implemented the two features separately, but that's not an excuse of a problem caused by the combined effect of the two features being introduced to the language.

**Reply by CyrusNajmabadi** [reply]
> @CyrusNajmabadi Please see my comment above on why these two orthogonal concepts are necessary for a useful implementation of immutability and therefore of records.

I did read it.  It again seems to be conflating things for me.  Not-null or null-ness is orthogonal to initialized or not.  

> Currently without "required members"

Requirement is a different concern.  And it's one we're aware of and working on proposals around.  But i don't see why things should be conflated here.  If something is required, there should be a way to say that.  It's not a facet of if it's null or not.  

> but that's not an excuse of a problem caused by the combined effect of the two features being introduced to the language.

I think you view the language too statically.  We intend for it to cahnge and develop over time, and move in areas where improved expressiveness can provide large amounts of value to the community.  In no way was 'records' ever a: "here it is, it's done now, it will never be anything but this" feature.  Instead, it was a simple way to express value types, with value-concerns, orthogonally to the work done to provide compiler warnings around null-tracking.

Records v1 was not intended to solve every use case.  Rather, we wanted to solve many, and then see what rose to the top of concerns at that point so that we could focus our efforts on the things most valuable.  We've already identified many areas for improvement, but all of them compliment what has come so far and does not seek to subvert or change the core idea of records here.

> but that's not an excuse of a problem 

I legitimately don't even see this as a problem.  Things make *more* sense to me here this way.  I would find it very confusing to behave otherwise.

**Reply by reinux** [reply]
> I did read it. It again seems to be conflating things for me.

I'm recognizing that a feature requires a confluence of two factors to work meaningfully. I'm not sure how you only recognize it as a conflation of those factors.

How well would `foreach` have worked if it hadn't been built with `IEnumerable` in mind? One's a language feature and the other's technically a BCL interface, but if those two things hadn't been planned in conjunction, `foreach` would have only been useful for arrays.

It's not uncommon for two orthogonal features to have to be planned together in order to achieve a goal. Surely coordination can't have broken down that severely.

> It's not a facet of if it's null or not.

I can't figure out how that would be useful in practical use. The reason for making a field required at initialization is that there is no acceptable default value. If a field is nullable, null _must_ be a valid state, so it would have to be an acceptable default value as well.

> I think you view the language too statically. 

What ever happened to measure twice, cut once?

I'm paraphrasing, but Anders Hejlsberg was quoted as saying that all languages become messier over time as new features are added, and one of the goals of language design is to try to slow that down as much as possible, while still adding useful features.

You can only remove a feature if it's completely broken. Otherwise, it just breaks things around it as time goes on. You can add a `required` keyword, but you can't add a `unrequired` keyword anymore, even though that would be far more sensible.

> but all of them compliment what has come so far and does not seek to subvert or change the core idea of records here.

The whole point of a record is that _everything_ is required up front. Otherwise it's a less useful OOP class with trivial syntactic sugar for limited use cases -- not a useful immutable construct.

I don't think it's controversial to claim that that's a fundamental feature of a record. Every functional language that I can think of, even relatively impure ones like F#, have defined records as requiring all fields to have a valid value at initialization, whether that's a meaningful default or provided to a constructor. Otherwise it just doesn't add much value to what OOP already does.

If nullable fields don't need to be initialized, that's fine, if anything, that's a slight convenience. But the opposite, having nothing required by default, means that even just to get your code to compile without warnings, you have to rely on an as-of-yet _unreleased_ keyword.

And by the way, the compiler is justified in giving that warning, too. The problem is that the language doesn't facilitate fixing it in a meaningful way.

I mean, how is that not a glaring omission and an obvious code smell?

**Reply by acaly** [reply]
First sorry for being a little bit offensive, but I wouldn't agree with some of the statement.

> I legitimately don't even see this as a problem.

In my opinion, the `{ get; init; }` syntax is only useful when the class is declared such that this property should be initialized in a initializer list of the `new` expression. Yes, it can be set by constructor (as suggested by the warning CS8618), but the constructor does not need this `init;` accessor. Therefore, if I see something like `{ get; init; }`, I know it is something that I _may_ initialize in the `new` expression. Without required members, this literally means _may_, or _may not_. If I choose not to initialize it (which is allowed by the `{ get; init; }`), no one will initialize it and leave it as null. So `{ get; init; }` implies "may not be explicitly initialized", which then implies "may be null". Here the contradiction comes, I can also declare it as "never to be null" by using a non-nullable type. How should the language allow me to say something may be null, while at the same time say it can never be null?

So I believe, given the non-nullable types, the `{ get; init; }` feature should really be considered as incomplete until we have `{ get; required init; }`. Otherwise a valid declaration may be self-contradictory as I described. In this case, whether or not you or I think this is orthogonal isn't important. In any case, there is a missing feature that prevents this self-contradiction situation.

**Reply by HaloFour** [reply]
@acaly 

> Yes, it can be set by constructor

The thing is, as a consumer of a class you have no idea whether or not a value was set in a constructor.  Or whether that's an auto-property at all.  All you (and the compiler) can tell is that there is a property with an `init` accessor method.  There's nothing to indicate whether or not the consumer of that class should be required to specify any value.  The nullability of the type of the property doesn't indicate this either.  That is why the language team is looking into required properties, as a means to convey this information to consumers of the class.

**Reply by acaly** [reply]
@HaloFour 

I understand that currently there is no way to express such information in the binary, but shouldn't that be considered as a problem of (being non-static, the current version of) the `init` accessor? Why did everyone above keep saying "orthogonal" without mentioning that there is a problem and we are fixing it?

**Reply by theunrepentantgeek** [reply]
@acaly you wrote:

> the { get; init; } syntax is only useful when the class is declared such that this property **should** be initialized in a initializer list of the new expression.

Emphasis added by me.

This is incorrect - the **init** syntax indicates the property **may** be initialized, not that it **should** be.

The reason some people keep going back to the orthogonal argument is that others keep arguing that nullability implies requiredness, when it doesn't, by design. 

It doesn't seem to matter how many times people like @HaloFour and @CyrusNajmabadi keep repeating the same explanations, people keep coming back to the same misunderstanding. I really admire their patience.

**Reply by reinux** [reply]
> The reason some people keep going back to the orthogonal argument is that others keep arguing that nullability implies requiredness, when it doesn't, by design.

The point is that if the purpose is to implement records as it's understood by virtually every other functional language, then the design is wrong. It's only correct if the purpose is to provide a redundant language feature in search of a problem.

The fact that a very normal, intended use case results in unclearable warnings should make that clear.

**Reply by CyrusNajmabadi** [reply]
> The point is that if the purpose is to implement records as it's understood by virtually every other functional language

This was never the purpose.

**Reply by CyrusNajmabadi** [reply]
> It's only correct if the purpose is to provide a redundant language feature in search of a problem.

V1 is not and had never been intended to be there for an use cases. We've been up front about that from the beginning. We even flipped majorly here, starting with an already that tried to solve everything we could throw at it. This quickly became an complex mess. So we started breaking down the problem space into independent pieces that could compose appropriately. If your domain fits into one of those composition points, hooray.  If not, we did what we've always done. Collect feedback and data and use that to prioritize future work and inform designs. 

As has been stated *numerous* times so far, the concept of 'requiredness' was known, but not included in v1 as we weren't certain how important it would be. It very much is on the menu for the next version of c#, and we've already started evaluating proposals for such a concept and how it could work both for records and other types.

**Reply by CyrusNajmabadi** [reply]
> It's not uncommon for two orthogonal features to have to be planned together in order to achieve a goal. Surely coordination can't have broken down that severely.

Again, as I've mentioned, we are literally viewing this scenario differently. I don't view it as broken. It makes sense to me what's going on, and it's what I would expect and want here. 

Note: I'm also personally a fan of having the 'required' concept. I asked for it a long time ago as part of an API design that truly wanted to force these values to be provided. 

However, I view that as the solution to this doesn't, not some strange crossover where orthogonal features bleed into each other in a way that only then works I'm that way for that case and for no others.

**Reply by CyrusNajmabadi** [reply]
> What ever happened to measure twice, cut once?

There's no pleasing everyone. We could have continued to punt records release after release. And then you'd have the people saying: stop trying to solve every use case, help me out today. 

We made a conscious and informed decision that the design was in no means 'done' but that we hit enough important use cases to release something worthwhile. We already had given a lot of time and thought into the extensions we felt could be important in the future, and we felt we would be and to provide in a manner that could be added later. 

I get that for you the particular areas we pushed out in happen to be really important. That happens. And it was going to happen as long as we didn't address every possible domain for 'records' (and trust me, there are a lot).

**Reply by CyrusNajmabadi** [reply]
> I'm paraphrasing, but Anders Hejlsberg

One is the best things I've learned over the years working with Anders is that it's far better to decompose these problems into lots of smaller and simpler solutions that compose, versus trying to add a lot of one-off complexity into a particular feature (esp with how it interacts with another). 

While I do not want to speak for Anders, I would say that he had a strong and direct influence on how this feature landed in the way it did. 

Furthermore, I'm getting very good that a lot of what came from records, that we've now gotten a ton of experience with, can now extend it to the rest of the language.

**Reply by acaly** [reply]
@theunrepentantgeek 

That sentence is my mistake. I should use _may_ in that sentence. However, besides this word, what I said is correct. If you keep reading, I did say that it is a _may_ instead of _should_:

> Therefore, if I see something like { get; init; }, I know it is something that I _may_ initialize in the new expression.

What I was trying to express is probably a little bit different than @reinux. My opinion is that this is not related with records. The `{ get; init; }` syntax itself is broken (or at least incomplete) on non-nullable types. The reason is, as you have also said, the API consumer may initialize it, so he/she may not. This declaration allows the consumer to not initialize those properties, which violates the nullable contract.

> others keep arguing that nullability implies requiredness, when it doesn't, by design

You are right. Non-nullability may not imply requiredness. Instead of leaving uninitialized properties (or variables) null, the language or runtime can instead implicitly set the non-nullable properties to some internally created objects, which throw something like a `ObjectNotInitializedException` upon accessed. Of course this will have its overhead, but in that case, initialization and nullables are orthogonal, because non-initialized properties are no longer null. A nullable property may be initialized or may be not initialized and a non-nullable property may also be initialized or may be not initialized. This is what I think is orthogonal. However, that is not the actual case. The current implementation always leaves the property null if not initialized. So initialization and nullables are not completely unrelated. If **by design** properties with non-nullable types are allowed to be left uninitialized, it is a bad design. 

Fortunately, C# isn't designed as you described. It does know non-nullable properties should be initialized. The warning CS8618 mentioned in the original post is generated exactly for this reason. The problem of this warning is not that it should not be generated. It's that it should further understand `{ get; init; }` and non-nullable type conflict with each other. Here what it asks the programmer to do is to make a constructor and assign some value to the property in the constructor, but that is not what `{ get; init; }` intends to do, which is to defer the initialization to `new` expression. As @HaloFour pointed out, forcing this deferral is impossible unless we have the `required init;` to replace the `init;`.

**Reply by acaly** [reply]
The reason I don't like the word "orthogonal" is because I think it only tries to hide potential problems from people. It seems that if you explained that A and B are orthogonal, we shouldn't ask any question that is related to both A and B. Why? Because A and B are orthogonal 🤣

**Reply by HaloFour** [reply]
@acaly 

> This declaration allows the consumer to not initialize those properties, which violates the nullable contract.

The nullable contract states that you can't set the value to null.  If you don't set a value you're not violating the contract.  I understand that this seemingly breaks the spirit of the contract, and I agree with that sentiment.

> The reason I don't like the word "orthogonal" is because I think it only tries to hide potential problems from people.

It's not about hiding problems.  It's about how solutions are related to one another.  `init` isn't intended to interact with NRTs and the combination isn't supposed to imply additional requirements.  That's where another also orthogonal feature will come in.  Those features, despite being orthogonal, can be combined to interact in a way to solve problems such as this.  It's like LINQ.  It looks like one giant feature, but it's really the combination of something like a half-dozen separate orthogonal features combined in an interesting way.

I also want to come out and state that I agree that POCOs + NRTs are "broken", in that the experience is quite poor, and that I do hope that the language addresses this soon.

**Reply by acaly** [reply]
@HaloFour Thanks for the explanation again. I think I might misunderstood some of the statements above. So by "orthogonal" you basically mean "when we designed records, we were so careful not to break things so we did not want to touch nullables and let's think about how these two interact later". It's not that the language team don't have thinking about it, but that they have not been addressed yet. However, what I interpreted from some of the comments above, for example, is different:

> I don't see the connection you're making here. You are conflating null vs initialized. I think it's a mistake to do that. Null can be totally legal for something. And whether or not something is initialized is an orthogonal concept.

It is like "these two features do not and will never touch each other and so don't ask any more".

**Reply by jnm2** [reply]
I don't think the point was "don't ask the question." The point was "here is the answer." Orthogonal would mean that all combinations of the two states of the two features can be useful (required and nullable, required and non-nullable, non-required and nullable, non-required and non-nullable). Since the "required and nullable" case exists and is useful, it would not make sense to tie the concept of being required to the concept of being non-nullable.

**Reply by acaly** [reply]
Are you saying that since required, nullable property is useful, we should give a warning to anyone using required, non-nullable properties? Does that make any sense?

**Reply by jnm2** [reply]
I'm not making the point but restating someone else's, but I'll give it a shot. I think the point is that since all four combos are useful, non-nullability should not imply requiredness, and nullability warnings for uninitialized, nonrequired, non-nullable members are just as valid and necessary as before.

**Reply by acaly** [reply]
No. You are again mixing required and initialized. Yes, requiredness and nullable types are orthogonal, but requiredness isn't part of the original question. If there were required feature, the author would use it for that property and the compiler should no longer give that warning at declaration site. The issue here is there is no such thing as required yet. What they claimed was that nullables and initialized are orthogonal, but you should never have something to be non-nullable and non-initialized for too long.

**Reply by jnm2** [reply]
> You are again mixing

I'm not sure what 'again' means since this was my first try, so I'll bow out because I can see I'm just confusing the issue. What I came to say was that I don't think "orthogonal" means "stop asking questions that relate these two things."

**Reply by acaly** [reply]
Sorry I did not mean you, but I have mentioned the difference of this already. https://github.com/dotnet/csharplang/discussions/4857#discussioncomment-1000777

The point is, @CyrusNajmabadi claiming nullables and initialized are orthogonal is just wrong. If they are orthogonal, the compiler won't have that warning. The author did not make a mistake of "conflating" things together by asking this question.

**Comment by theunrepentantgeek** [comment]
> It is like "these two features do not and will never touch each other and so don't ask any more".

If that's the meaning you took, then I expressed myself poorly. 

What I've been trying to say is "These two features are not the same feature, and so please stop asking for them to be the same feature."

@acaly, you wrote:

> Instead of leaving uninitialized properties (or variables) null, the language or runtime can instead implicitly set the non-nullable properties to some internally created objects, which throw something like a ObjectNotInitializedException upon accessed. 

This reads to me as trying to get the compiler to enforce some kind of requiredness _as a part of the nullability feature_ - trying to bring in requiredness by stealth. 

The _nullability_ (or not) of a property **cannot** be used to reliably infer the _requiredness_ (or not) of that property.

I understand that you have a very common scenario where nullability and requiredness are very closely related. There are a lot of people who have that scenario, and this is why the C# LDM are designing/implementing requiredness. 

But, and this is important: There are valid, useful, and common scenarios where a non-nullable property is not required.

The CS8618 warning is correct - and like every other warning in C#, it's up to the developer to decide _whether_ to fix it (warnings are, after all, not errors), and further _how_ to fix it.

The warning is telling the developer that there's something missing here. _Either_ the constructor needs to be modified to assign a value, _or_ it needs to be assigned during initialization, _or_ the warning can be suppressed, _or_ the warnings can be ignored.

**Reply by theunrepentantgeek** [reply]
Let me expand on this:

> There are valid, useful, and common scenarios where a non-nullable property is not required.

Here's an example. First, the class declaration:

``` csharp
public class NotificationSender
{
    public INotificationStyle Style { get; required init; }
    public IRetryPolicy? RetryPolicy { get; init; }

    public NotificationSender() 
    {
        RetryPolicy = new ExponentialRetryPolicy();
    }
}
```

Now, initialization:

``` csharp
var sender = new NotificationSender()
{
    Style = new MessengerStyle()
};
```

The property 'Style' is required, because the `NotificationSender` can't know which style is needed. 

The property `RetryPolicy` is nullable, because it is valid to have no retry policy - it just means no retries. The constructor sets a useful default (the `ExponentialRetryPolicy`), but the class design permits this to be overridden. Consumers _may_ choose to do things differently, but they are not _required_ to do so.

**Reply by acaly** [reply]
OK, but the original post asks for a thinking behind the design. First, that is not to say they are the same feature. Second, it does not answer the question of "what is the thinking behind A and B" by saying "A and B are two features".

I agree that non-nullable, non-required property has its use case. The problem is, then we have to write a constructor and set to it some valid value. If no valid value is known, it has to use a constructor parameter, which eliminates all the advantages of init accessor. (By the way, your example does not have a non-nullable, non-required property.)

**Reply by acaly** [reply]
> The nullability (or not) of a property cannot be used to reliably infer the requiredness (or not) of that property.

The are 2 different levels of requiredness: (1) whether the property or variable should be initialized, (2) whether the property or variable should be initialized externally by consumer.

For (1) we (and the compiler) can infer from non-nullability in that non-nullables must be initialized somewhere. They are related. For (2), I agree that, we can't infer (2) from non-nullability, but we don't have this concept of (2) yet in the language. So you can't really say they are totally unrelated until we have (2).

**Reply by DavidArno** [reply]
@theunrepentantgeek,

> ... The warning is telling the developer that there's something missing here ... 

And that thing that is missing is `required`. There are too many here that have simply rushed to declare records and init properties as being orthogonal to each other when in reality they are clearly interrelated. There are four combinations of nullable/optional for properties and for one of them (non nullable and mandatory) not only is it unenforceable, but it gives a warning that has to either be ignored or suppressed.

I think there is often too much of a rush to declare the question invalid and to shut the questioner down in discussions here. And this question is a classic example of that happening. The answer to the question of why does the problem of mixing NRTs and records exist isn't "they are unrelated; there isn't a problem" and it's deeply unhelpful that some offered that as a response, especially since it is a false response. The truth is - for good or bad - the decision was made to partially develop record functionality and that left this functional hole. In time it be filled by `required` and the problem will go away, but it simply has to be worked around in the meantime. Sometimes just being open and honest is the simplest solution.

**Reply by jnm2** [reply]
I think programmers are particularly susceptible to overusing and abusing pet jargon like "orthogonal" or "conflate" and it can have a cold effect on the conversation. I see this even when the term is being used properly and people aren't just leaping to it too soon as a conversational move that feels good.

**Reply by jnm2** [reply]
> The are 2 different levels of requiredness: (1) whether the property or variable should be initialized, (2) whether the property or > variable should be initialized externally by consumer.
> 
> For (1) we (and the compiler) can infer from non-nullability in that non-nullables must be initialized somewhere.

@acaly I think I see a clarification. If the compiler were to infer your "level 1" requiredness but then elevate it to "level 2," that would be an API change that should be annotated and locked in explicitly using a keyword like `required` so that your API doesn't fluctuate implicitly based on indirect changes to the private implementation. When that keyword doesn't exist, the compiler still shouldn't elevate it to "level 2," but it would be a foot gun for it to neglect to warn that there might be a null in the field marked as never null. I see only one option for the compiler when a `required` keyword is not used (or doesn't yet exist), and that is the current behavior.
